src/uct/ib/rc/verbs/rc_verbs_ep.c

```diff
@@ -15,11 +15,76 @@
 #include <ucs/arch/bitops.h>
 #include <uct/ib/base/ib_log.h>
 
+#define UCT_RC_VERBS_CHECK_FENCE(_iface, _ep) \
+    do { \
+        ucs_status_t _stat; \
+        _stat = uct_rc_verbs_ep_is_fence(_iface, _ep); \
+        if (_stat != UCS_OK) { \
+            return _stat; \
+        } \
+    } while (0)
+
+#define UCT_RC_VERBS_CHECK_RES(_iface, _ep) \
+    UCT_RC_CHECK_RES(&(_iface)->super, &(_ep)->super) \
+    UCT_RC_VERBS_CHECK_FENCE(_iface, _ep)
+
+#define UCT_RC_VERBS_CHECK_RES_FC(_iface, _ep) \
+    UCT_RC_CHECK_RES(&(_iface)->super, &(_ep)->super) \
+    UCT_RC_CHECK_FC(&iface->super, &ep->super, id) \
+    UCT_RC_VERBS_CHECK_FENCE(_iface, _ep)
+
 void uct_rc_verbs_txcnt_init(uct_rc_verbs_txcnt_t *txcnt)
 {
     txcnt->pi = txcnt->ci = 0;
 }
 
+static UCS_F_ALWAYS_INLINE int
+uct_rc_verbs_ep_fence_state(uct_rc_verbs_iface_t *iface, uct_rc_verbs_ep_t *ep)
+{
+    return iface->super.tx.fi.fence_beat - ep->fi.fence_beat;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_rc_verbs_ep_is_fence(uct_rc_verbs_iface_t *iface, uct_rc_verbs_ep_t *ep)
+{
+    ucs_status_t status;
+
+    if (iface->super.config.fence_mode != UCT_RC_FENCE_MODE_STRONG) {
+        return UCS_OK;
+    }
+
+    if (!uct_rc_ep_is_fence(&iface->super, &ep->fi)) {
+        return UCS_OK;
+    }
+    
+    if (uct_rc_verbs_ep_fence_state(iface, ep) == UCT_RC_FENCE_INPROGRESS) {
```

Comment:
so the diff between ep fence_beat and iface fence_beat is the fence state?
what happens when we do fence on ep and not on iface??